### DISCUSSION TITLE
Dynamic wms layer and customise component

### DISCUSSION THREAD

**Discussion by samareshdasglobal** [TerriaJS/terriajs#7049] [Q&A]
Hello,
I've had a positive experience working with Terrimap and its functionality. Now, I aim to integrate TerriaJS into another React web application. In Terriamap, I observed that the WMS layers' URLs are configured in the configuration file. However, in my application, the URL needs to be dynamic, varying based on the logged-in user and selected product. It should change the URL for WMS capability requests on the fly. Additionally, I require the ability to filter layers for display in the UI. Can I customize these functionalities while using this library?

From my observations, all components seem to be developed within the library. If I need to customize certain components, how can I achieve this while maintaining the integrity of the library?

**Comment by nf-s** [comment]
Hi @samareshdasglobal 

TerriaMap/TerriaJS isn't really designed for that level of customization - for drastic changes to the UI, it would be possible to add new components to TerriaMap in order to keep TerriaJS as a dependency, but it would require a lot of work as basically all components are within the TerriaJS repository. For example, you would probably need to move `lib/ReactViews/StandardUserInterface/StandardUserInterface.tsx` into TerriaMap.

If you are able to restrict your changes to the model layer, and keep existing UI - you can extend TerriaJS catalog items in TerriaMap - for example create a new `CustomWebMapServiceCatalogItem` that extends `terriajs/lib/Models/Catalog/Ows/WebMapServiceCatalogItem`, and then add hooks to change the URL based on user/selected product.

It is also possible to make simple changes to the UI in TerriaMap - for example adding extra screen elements
